Fixed #29984 -- Added QuerySet.iterator() support for prefetching related objects. (#15334)
django/django: docs/ref/models/querysets.txt

Note that using ``iterator()`` on a ``QuerySet`` which has already been
evaluated will force it to evaluate again, repeating the query.

.. versionchanged:: 4.1

    ``iterator()`` is compatible with previous calls to ``prefetch_related()``
    as long as ``chunk_size`` is given. Larger values will necessitate fewer
    queries to accomplish the prefetching at the cost of greater memory usage.

>     A value under 1000 should be used on SQLite.

Author: Thanks, Simon, that's smart. I gave it a shot, let me know how it sounds. I stopped short of a full code example.

Note that using ``iterator()`` on a ``QuerySet`` which has already been
evaluated will force it to evaluate again, repeating the query.

.. versionchanged:: 4.1

    ``iterator()`` is compatible with previous calls to ``prefetch_related()``
    as long as ``chunk_size`` is given. Larger values will necessitate fewer
    queries to accomplish the prefetching at the cost of greater memory usage.

>     As the maximum number of terms in an SQL ``IN`` clause on SQLite is 999,

Reviewer: Perhaps this text is fine as-is, given it is talking about the lowest common denominator supported by Django, but AFAIK recent versions of SQLite (`3.32+`) aren't encumbered by this, so _one_ day when that's the minimum supported version, it won't be relevant (although maybe that day is 'today' if the `DatabaseFeatures` were modified?)

Reviewer 2: I would leave `999` because `SQLite` has other limits that can cause a crash, e.g. `SQLITE_MAX_COLUMN` (see f91b5a7e4b2ac436b1fe3065b23217543d0f54c1 and ticket-26063), or `SQLITE_MAX_LENGTH`.

Reviewer 2: Also, the maximum number of terms is also limited on Oracle by `2^16 - 1`. Maybe we could use a more general note, e.g.
```
    On some databases (e.g. Oracle,
    `SQLite <https://www.sqlite.org/limits.html#max_variable_number>`_), the
    maximum number of terms in an SQL ``IN`` clause might be limited. Hence
    values below this limit should be used. (In particular, when prefetching
    across two or more relations, a ``chunk_size`` should be small enough that
    the anticipated number of results for each prefetched relation still falls
    below the limit.)
```

Note that using ``iterator()`` on a ``QuerySet`` which has already been
evaluated will force it to evaluate again, repeating the query.

``iterator()`` is compatible with previous calls to ``prefetch_related()`` as
long as ``chunk_size`` is given. Larger values will necessitate fewer queries
to accomplish the prefetching at the cost of greater memory usage.

On some databases (e.g. Oracle, `SQLite
<https://www.sqlite.org/limits.html#max_variable_number>`_), the maximum number
of terms in an SQL ``IN`` clause might be limited. Hence values below this
limit should be used. (In particular, when prefetching across two or more
relations, a ``chunk_size`` should be small enough that the anticipated number
of results for each prefetched relation still falls below the limit.)

So long as the QuerySet does not prefetch any related objects, providing no

> value for ``chunk_size`` will result Django using an implicit default of 2000.

Author: ```suggestion
value for ``chunk_size`` will result in Django using an implicit default of 2000.
```

Depending on the database backend, query results will either be loaded all at
once or streamed from the database using server-side cursors.

.. versionchanged:: 4.1

    Support for prefetching related objects was added.

.. deprecated:: 4.1

    Using ``iterator()`` on a queryset that prefetches related objects without

>     providing the ``chunk_size`` is deprecated. In Django 5.0, the exception
>     will be raise.

Author: ```suggestion
    providing the ``chunk_size`` is deprecated. In Django 5.0, an exception
    will be raised.
```

the database driver. Larger batches decrease the overhead of communicating with
the database driver at the expense of a slight increase in memory consumption.

So long as the QuerySet does not prefetch any related objects, providing no

> value for ``chunk_size`` will result Django using an implicit default of 2000,

Author: ```suggestion
value for ``chunk_size`` will result in Django using an implicit default of 2000,
```
